src/Handler/StreamHandler.php

```diff
@@ -308,9 +308,17 @@ function () use ($context, $params) {
         );
 
         return $this->createResource(
-            function () use ($request, &$http_response_header, $context) {
+            function () use ($request, &$http_response_header, $context, $options) {
                 $resource = fopen((string) $request->getUri()->withFragment(''), 'r', null, $context);
                 $this->lastHeaders = $http_response_header;
+
+                if (isset($options['read_timeout'])) {
+                    $readTimeout = $options['read_timeout'];
+                    $sec = (int) $readTimeout;
+                    $usec = ($readTimeout - $sec) * 100000;
+                    stream_set_timeout($resource, $sec, $usec);
```

Comment:
To be honest I wonder if it would be better to pass the timeout to `stream_context_create`. It supports float values so it wouldn't require the magic above. One could even argue that the same functionality is already available via the undocumented `stream_context` option. WDYT @arnaud-lb ?